component_sdk/python/kfp_component/_base_op.py

```diff
@@ -0,0 +1,213 @@
+            self._stage_states()
+            raise
+        finally:
+            signal.signal(signal.SIGTERM, original_sigterm_hanlder)
+
+    def on_executing(self):
+        """Triggers when execute method is called.
+        Subclass should override this method.
+        """
+        pass
+
+    def on_cancelling(self):
+        """Triggers when the operation should be cancelled.
+        Subclass should override this method.
+        """
+        pass
+
+    def _exit_gracefully(self, signum, frame):
+        logging.info('SIGTERM signal received.')
+        if self._should_cancel():
+            self.on_cancelling()
+        else:
+            self._stage_states()
+
+    def _should_cancel(self):
+        """Checks argo's execution config deadline and decide whether the operation
+        should be cancelled.
+
+        Argo cancels workflow by setting deadline to 0 and sends SIGTERM
+        signal to main container with 10s graceful period.
+        """
+        pod = self._load_pod()
+        if not pod or not pod.metadata or not pod.metadata.annotations:
+            return False
+        
+        argo_execution_config_json = pod.metadata.annotations.get('workflows.argoproj.io/execution', None)
+        if not argo_execution_config_json:
+            return False
+        
+        try:
+            argo_execution_config = json.loads(argo_execution_config_json)
+        except Exception as e:
+            logging.error("Error deserializing argo execution config: {}".format(e))
+            return False
+        
+        deadline_json = argo_execution_config.get('deadline', None)
+        if not deadline_json:
+            return False
+        
+        try:
+            deadline = datetime.strptime(deadline_json, '%Y-%m-%dT%H:%M:%SZ')
+        except Exception as e:
+            logging.error("Error converting deadline string to datetime: {}".format(e))
+            return False
+        
+        return datetime.now() > deadline
+
+    def _stage_states(self):
+        if not self._staging_bucket or not self._staging_path or not self.staging_states:
+            return
+
+        try:
+            bucket = self._storage.get_bucket(self._staging_bucket)
+        except google.cloud.exceptions.NotFound:
+            logging.error('Unable to find staging bucket {}.'.format(self._staging_bucket))
+            return
+ 
+        blob = bucket.blob(self._staging_path)
+
+        try:
+            states_json = json.dumps(self.staging_states)
+        except TypeError as e:
+            logging.error('Unable to dump staging states: {}'.format(self.staging_states))
+            return
+
+        try:
+            blob.upload_from_string(states_json)
+        except google.cloud.exceptions.GoogleCloudError as e:
+            logging.error('Failed to upload staging states: {}'.format(e))
+            return
+
+    def _load_k8s_client(self):
+        config.load_incluster_config()
```

Comment:
Why does the container needs the cluster config? Does this require privilleges that we may not want to grant to individual containers in the future?

Reply:
It is required before reading the pod metadata. Without the permission, argo sidecar won't work. I don't think we can remove the privileges unless we replace argo.

Reply:
SG, thanks.